Sources/ComposableArchitecture/SwiftUI/Binding.swift

```diff
@@ -114,39 +177,135 @@ extension BindableAction {
+    _ keyPath: WritableKeyPath<State, BindingState<Value>>,
     _ value: Value
   ) -> Self {
     self.binding(.set(keyPath, value))
   }
 }
 
-extension ViewStore where ViewAction: BindableAction, ViewAction.State == ViewState {
-  /// Returns a binding to the resulting bindable state of a given key path.
-  ///
-  /// - Parameter keyPath: A key path to a specific bindable state.
-  /// - Returns: A new binding.
-  public func binding<Value: Equatable>(
-    _ keyPath: WritableKeyPath<ViewState, BindableState<Value>>,
-    file: StaticString = #file,
-    fileID: StaticString = #fileID,
-    line: UInt = #line
-  ) -> Binding<Value> {
-    self.binding(
+/// A ``BindingState`` variant that is suitable for `ViewState`s.
+///
+/// You can only build these values using dynamic member lookup of ``BindingState`` properties on
+/// the ``BindingViewStates`` value from ``BindableStateProtocol/bindings``.
+///
+/// Because you're defining the ``BindingViewState`` value directly, you must assign the value
+/// of the property wrapper private storage using the underscored property name:
+///
+/// ```swift
+/// struct ViewState: Equatable {
+///   @BindingViewState var text: String
+///   init(state: State) {
+///     self._text = state.bindings.$text
+///   }
+/// }
+/// ```
+///
+/// Read <doc:Bindings> for more information.
+@propertyWrapper
+public struct BindingViewState<Value> {
+  let keyPath: AnyKeyPath
+  public var wrappedValue: Value
+  public var projectedValue: Self { self }
+  internal init<State>(
+    wrappedValue: Value,
+    keyPath: WritableKeyPath<State, BindingState<Value>>
+  ) {
+    self.wrappedValue = wrappedValue
+    self.keyPath = keyPath
+  }
+}
+extension BindingViewState: Equatable where Value: Equatable {}
+
+extension BindingViewState: Hashable where Value: Hashable {}
+
+extension BindingViewState: CustomReflectable {
+  public var customMirror: Mirror {
+    Mirror(reflecting: self.wrappedValue)
+  }
+}
+
+extension BindingViewState: CustomDumpRepresentable {
+  public var customDumpValue: Any {
+    self.wrappedValue
+  }
+}
+
+extension BindingViewState: CustomDebugStringConvertible
+where Value: CustomDebugStringConvertible {
+  public var debugDescription: String {
+    self.wrappedValue.debugDescription
+  }
+}
+
+// `BindingViewState` dynamic member lookup.
+extension ViewStore where ViewAction: BindableAction {
+  public subscript<Value>(dynamicMember keyPath: KeyPath<ViewState, BindingViewState<Value>>)
+    -> Binding<Value> where Value: Equatable
+  {
+    let stateKeyPath =
+      self.state[keyPath: keyPath].keyPath
+      as! WritableKeyPath<ViewAction.State, BindingState<Value>>
+    return self.binding(
+      get: { $0[keyPath: keyPath].wrappedValue },
+      send: { value in
+        #if DEBUG
+          let debugger = BindableActionViewStoreDebugger(
+            value: value,
+            bindableActionType: ViewAction.self,
+            // TODO: Restore context
```

Comment:
Should we try to propagate at least some context to here? We could have `BindingState` capture the file/fileID/line at the moment of creation to tie the warning to the field on the `State` that was used incorrectly.

Same for the TODO down below.

Reply:
<img width="738" alt="Capture d’écran 2023-01-04 à 20 49 41" src="https://user-images.githubusercontent.com/35562418/210671361-bfdef18d-de74-4f50-bbcf-5853e18b4c18.png">
Do you mean something like this?

Reply:
Yeah maybe, but I think it needs finessing. I don't think it's correct to mention `Binding.swift:357` at all since it's library code that isn't helpful. So the whole message may need to be rethought.

Reply:
I'm capturing each property wrapper's `init` context, and I'm also passing the one from `BindingState` to `BindingViewState` (so this one has both at hand).
In both cases, I'm pointing one the `BindingState` value in the runtime warning.
I just pushed a commit, but we can reverse if this is not what you had in mind.

Reply:
Ok, I just saw your comment. I'll try to improve this.